### PR TITLE
Improved stacking feature and added unnest into stack dialog

### LINKED ISSUE #6687: Add dialogue or radio button for the unnest tokens function
This is all part of text mining.  The function is in the tidytext package, which is (I think) already in R-Instat.  The use is described in the [vignette](https://mran.microsoft.com/snapshot/2018-08-31/web/packages/tidytext/vignettes/tidytext.html).  In addition there is (yet another) interesting book that is available as a soft copy.  It covers the same materials and more in more detail.  It is called [Text Mining with R: A Tidy Approach.](https://www.tidytextmining.com/tfidf.html)

This function/dialogue takes texts and breaks it up into words.  Usually each row has a line of text, or perhaps a sentence or a paragraph.  Here is an example:
image

Then the function `unnest_tokens(jane,word, text)`  takes the variable called text, and produces a new data frame with a variable called word.

image

So this is like the current stack, except it is stacking the components of a single (quite complicated) variable.

So, my suggestion is to change the name of Stack to Stack (Gather).  Then to change the existing dialogue (at the same time) to use Gather.  There are 2 buttons.  One is Gather and that does what the reshape2::melt function does.  But we change it to use tidyR and gather.  Then we add another radio button called unnest and use that for the unnest_tokens function.

a) It has the usual data selector and a single receiver for the column with the texts.  This will (I assume be a character variable.
b) Then there is (at the bottom.the Stack Data Into is just Data into (and the default is word, but see below.
c) There is (like stack), new data frame name control.
d) Below the single receiver there is a label `Token`: with a pull-down - also into which you can type..  The default is words and the list is given in the command.  
e) Then label Format with pull-down and options given in the function.  Default is text.
f) Below that is a checkbox, default unchecked with label `to Lower Case`
g) Label Pattern/Option with a field into which you can type:  This gives pattern for regex or for use-defined function. There is a tooltip with Example "Chapter [\\d]" for regex token or " " for user defined token. 
h) Label Strip on left then checkbox default unchecked Punctuation another Non-Alphanumeric another Numeric,  another URL.  If checked then strip_punct = TRUE, strip_non_alphanum = TRUE, strip_numeric = TRUE, strip_url = TRUE

### PR DESCRIPTION
**Pull request by Vitalis95**
Fixes #6687  @rdstern @shadrackkibet @lloyddewit This is now ready for review.

### CODE REVIEW

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 this is great.  It works already!  I am not used to that.
One minor change is to restrict the single variable to just receive character-type variables.  I tried with others and it gives an error.  Then I think it can be merged.

**Comment by Vitalis95** [comment]
@rdstern  could you have a look at this again

**Comment by rdstern** [comment]
Thanks @Vitalis95 - will do

**Review by rdstern — DISMISSED** [review]
Looks very good!
So I have approved.  Let's see what @lloyddewit or @shadrackkibet thinks.  It would be great for this to be merged soon.

File: instat/dlgStack.vb [hunk 1]
```diff
@@ -37,50 +38,110 @@ Public Class dlgStack
...
+        dctToken.Add("characters", Chr(34) & " characters" & (34))
+        dctToken.Add("character_shingles", Chr(34) & "character_shingles" & Chr(34))
+        dctToken.Add("ngrams", Chr(34) & "ngrams" & Chr(34))
+        dctToken.Add("skip_ngrams", Chr(34) & "skip_ngrams" & Chr(34))
+        dctToken.Add("sentences", Chr(34) & "sentences" & Chr(34))
+        dctToken.Add("lines", Chr(34) & "lines" & Chr(34))
+        dctToken.Add("paragraphs", Chr(34) & "paragraphs" & Chr(34))
+        dctToken.Add("regex", Chr(34) & "regex" & Chr(34))
+        dctToken.Add("tweets", Chr(34) & "tweets" & Chr(34))
+        dctToken.Add("ptb", Chr(34) & "ptb" & Chr(34))
```
Comment (shadrackkibet) [inline instat/dlgStack.vb:86]: Use `Penn treebank` as the key here.

File: instat/dlgStack.vb [hunk 2]
```diff
@@ -37,50 +38,110 @@ Public Class dlgStack
...
+
+        ucrInputValuesTo.SetParameter(New RParameter(" values_to", 4))
+        ucrInputValuesTo.SetRDefault(Chr(34) & "value" & Chr(34))
+        ucrInputValuesTo.SetLinkedDisplayControl(lblValuesTo)
+
+        ucrInputToken.SetLinkedDisplayControl(lblToken)
+        ucrInputToken.SetParameter(New RParameter("token", 3))
+        dctToken.Add("words", Chr(34) & "words" & Chr(34))
+        dctToken.Add("characters", Chr(34) & " characters" & (34))
+        dctToken.Add("character_shingles", Chr(34) & "character_shingles" & Chr(34))
```
Comment (shadrackkibet) [inline instat/dlgStack.vb:78]: Use `Character shingles` as a key. Also, make sure the first letter is CAPITALISED in all cases.

File: instat/dlgStack.vb [hunk 3]
```diff
@@ -37,50 +38,110 @@ Public Class dlgStack
...
+        ucrInputToken.SetItems(dctToken)
+        ucrInputToken.SetRDefault(Chr(34) & "words" & Chr(34))
+
+        ucrInputFormat.SetParameter(New RParameter("format", 4))
+        ucrInputFormat.SetLinkedDisplayControl(lblFormat)
+        dctFormat.Add("text", Chr(34) & "text" & Chr(34))
+        dctFormat.Add("man", Chr(34) & "man" & Chr(34))
+        dctFormat.Add("latex", Chr(34) & "latex" & Chr(34))
+        dctFormat.Add("html", Chr(34) & "html" & Chr(34))
+        dctFormat.Add("xml", Chr(34) & "xml" & Chr(34))
```
Comment (shadrackkibet) [inline instat/dlgStack.vb:96]: Same case here

File: instat/dlgStack.vb [hunk 4]
```diff
@@ -181,47 +219,62 @@ Public Class dlgStack
...
-        If ucrChkStackMultipleSets.Checked Then
-            ucrBase.clsRsyntax.SetBaseRFunction(clsReshapeFunction)
-            ucrReceiverColumnsToBeStack.SetMeAsReceiver()
-            'This is needed in case the value in ucrFactorInto is the R default for melt ("variable")
-            'in which case it would have been removed from all functions (R default shared by all functions)
-            If Not ucrFactorInto.IsEmpty Then
-                clsReshapeFunction.AddParameter("timevar", Chr(34) & ucrFactorInto.GetText() & Chr(34))
+    Private Sub ucrChkCarryColumns_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkCarryColumns.ControlValueChanged
+        If rdoPivotLonger.Checked Then
+            If ucrChkCarryColumns.Checked Then
```
Comment (shadrackkibet) [inline instat/dlgStack.vb:267]: `AndAlso`?

File: instat/dlgStack.vb [hunk 5]
```diff
@@ -181,47 +219,62 @@ Public Class dlgStack
...
+            If Not ucrReceiverColumnsToCarry.GetVariableNamesAsList.Contains(strColumn) Then
+                clsSelectFunction.AddParameter(strColumn, strColumn, iPosition:=iPosition, bIncludeArgumentName:=False)
+                iPosition = iPosition + 1
+            End If
+        Next
+        For Each strColumn In ucrReceiverColumnsToCarry.GetVariableNamesAsList
+            If Not ucrReceiverColumnsToBeStack.GetVariableNamesAsList.Contains(strColumn) Then
+                clsSelectFunction.AddParameter(strColumn, strColumn, iPosition:=iPosition, bIncludeArgumentName:=False)
+                iPosition = iPosition + 1
+            End If
```
Comment (shadrackkibet) [inline instat/dlgStack.vb:245]: Do we really need a loop? This are just columns added into select function which could be done via setRcode.

File: instat/dlgStack.vb [hunk 6]
```diff
@@ -89,83 +150,60 @@ Public Class dlgStack
...
-        clsReshapeFunction.SetRCommand("reshape")
-        clsReshapeFunction.SetAssignTo(ucrSelectorStack.ucrAvailableDataFrames.cboAvailableDataFrames.Text & "_stacked", strTempDataframe:=ucrSelectorStack.ucrAvailableDataFrames.cboAvailableDataFrames.Text & "_stacked")
-        clsReshapeFunction.AddParameter("direction", Chr(34) & "long" & Chr(34))
-        clsReshapeFunction.AddParameter("timevar", Chr(34) & "variable" & Chr(34))
+        clsUnnestTokensFunction.SetPackageName("tidytext")
+        clsUnnestTokensFunction.SetRCommand("unnest_tokens")
 
-        clsSplitColumnsInGroups.SetRCommand("split_items_in_groups")
-        clsSplitColumnsInGroups.AddParameter("num", 2, iPosition:=1)
+        clsSelectFunction.SetRCommand("select")
```
Comment (shadrackkibet) [inline instat/dlgStack.vb:169]: `SetPackageName`?

File: instat/dlgStack.vb [hunk 7]
```diff
@@ -37,50 +38,110 @@ Public Class dlgStack
...
-        ucrSelectorStack.SetParameter(New RParameter("data", 0))
+        ucrSelectorStack.SetParameter(New RParameter("tbl", 0))
         ucrSelectorStack.SetParameterIsrfunction()
 
-        ucrColumnsToCarryReceiver.SetParameter(New RParameter("id.vars", 1))
-        ucrColumnsToCarryReceiver.Selector = ucrSelectorStack
-        ucrColumnsToCarryReceiver.SetParameterIsString()
-        ucrColumnsToCarryReceiver.bAddRemoveParameter = False
-        ucrColumnsToCarryReceiver.SetValuesToIgnore({"NULL"})
+        ucrReceiverColumnsToCarry.SetParameter(New RParameter("id.vars", 1))
```
Comment (shadrackkibet) [inline instat/dlgStack.vb:49]: Check the parameter name to use?

**Review by shadrackkibet — COMMENTED** [review]
@Vitalis95 this is a good start! I just have a few comments. Please work with @Ivanluv to resolve them. I will also be around to support you.

**Comment by Vitalis95** [comment]
@shadrackkibet please have a look at it.

File: instat/dlgStack.vb (hunk 5 above)
Comment (shadrackkibet) [reply instat/dlgStack.vb:244]: @Ivanluv if this sounds a bit complicated could you just make sure that these parameters are removed?

**Review by rdstern — APPROVED** [review]
Still seems fine to me

**Comment by lloyddewit** [comment]
@shadrackkibet I see that you already reviewed this PR - Thank you! Please could you also do the final review and merge? thanks

**Review by shadrackkibet — APPROVED** [review]

**Pull request merged**